test/functional/test_framework/miniscript.py

```diff
@@ -0,0 +1,1709 @@
+
+    def construct_hash256(self, hash_digest):
+        assert isinstance(hash_digest, bytes) and len(hash_digest) == 32
+        self._hash256 = hash_digest
+        self._construct(NodeType.HASH256, Property().from_string("Bonudm"),
+                        [],
+                        self._hash256_sat, self._hash256_dsat,
+                        [OP_SIZE, 32, OP_EQUALVERIFY,
+                         OP_HASH256, hash_digest, OP_EQUAL],
+                        'hash256('+hash_digest.hex()+')'
+                        )
+        return self
+
+    def construct_ripemd160(self, hash_digest):
+        assert isinstance(hash_digest, bytes) and len(hash_digest) == 20
+        self._ripemd160 = hash_digest
+        self._construct(NodeType.RIPEMD160, Property().from_string("Bonudm"),
+                        [],
+                        self._ripemd160_sat, self._ripemd160_dsat,
+                        [OP_SIZE, 32, OP_EQUALVERIFY,
+                         OP_RIPEMD160, hash_digest, OP_EQUAL],
+                        'ripemd160('+hash_digest.hex()+')'
+                        )
+        return self
+
+    def construct_hash160(self, hash_digest):
+        assert isinstance(hash_digest, bytes) and len(hash_digest) == 20
+        self._hash160 = hash_digest
+        self._construct(NodeType.HASH160, Property().from_string("Bonudm"),
+                        [],
+                        self._hash160_sat, self._hash160_dsat,
+                        [OP_SIZE, 32, OP_EQUALVERIFY,
+                         OP_HASH160, hash_digest, OP_EQUAL],
+                        'hash160('+hash_digest.hex()+')'
+                        )
+        return self
+
+    def construct_thresh_m(self, k, keys_n):
+        self._k = k
+        n = len(keys_n)
+        assert n >= k >= 1
+        prop_str = "Bnudems"
+        self._pk_n = [key.get_bytes() for key in keys_n]
+        desc = "thresh_m("+str(k)+","
+        for idx, key_b in enumerate(self._pk_n):
+            desc += key_b.hex()
+            desc += "," if idx != (n-1) else ")"
+        self._construct(NodeType.THRESH_M, Property().from_string(prop_str),
+                        [],
+                        self._thresh_m_sat, self._thresh_m_dsat,
+                        [k, *self._pk_n, n, OP_CHECKMULTISIG], desc
+                        )
+        return self
+
+    def construct_and_v(self, child_x, child_y):
+        prop_str = ""
+        prop_str += "B" if child_x.p.V and child_y.p.B else ""
+        prop_str += "K" if child_x.p.V and child_y.p.K else ""
+        prop_str += "V" if child_x.p.V and child_y.p.V else ""
+        prop_str += "u" if child_y.p.u else ""
+        prop_str += "n" if child_x.p.n or (child_x.p.z and child_y.p.n) else ""
+        prop_str += "z" if child_x.p.z and child_y.p.z else ""
+        prop_str += "o" if (child_x.p.z and child_y.p.o) or \
+            (child_x.p.o and child_y.p.z) else ""
+        prop_str += "f" if child_x.p.f or child_y.p.f else ""
```

Comment:
I think this should be ` prop_str += "f" if child_x.p.s or child_y.p.f else "" ` . Since the left child must have a base type`V`, it always has malleability property `f`. The above statement will make all `and_v` has property `f`. 
There is only one possible non-canonical dissat being [sat(X), dissat(Y)] which gives us self.f = `sx or fy`. 

One valid miniscript counter-example: `and_v(vc:pk(A),and_v(v:after(10),sha256(H)))` where right branch of first `and_v` is not an `f`.

Reply:
This was a mistake in encoding the property rules, thank you.